yarn-project/simulator/src/avm/journal/journal.ts

```diff
@@ -133,48 +150,41 @@ export class AvmPersistableStateManager {
     this.log.debug(
       `Storage read  (address=${contractAddress}, slot=${slot}): value=${value}, exists=${exists}, cached=${cached}`,
     );
-    const treeLeafSlot = computePublicDataTreeLeafSlot(contractAddress, slot);
-    const indexedLeafPreimage = await this.merkleTree.getLeafPreimage(
-      MerkleTreeId.PUBLIC_DATA_TREE,
-      treeLeafSlot.toBigInt(),
-    );
 
-    if (indexedLeafPreimage === undefined) {
-      const lowLeafIndex = await this.merkleTree.getPreviousValueIndex(MerkleTreeId.PUBLIC_DATA_TREE, slot.toBigInt());
-      const lowLeafPreimage = await this.merkleTree.getLeafPreimage(MerkleTreeId.PUBLIC_DATA_TREE, lowLeafIndex!.index);
-      const lowLeafPath = await this.merkleTree.getSiblingPath(MerkleTreeId.PUBLIC_DATA_TREE, lowLeafIndex!.index);
-      const storageLeafPreimage = new PublicDataTreeLeafPreimage(
-        slot,
-        value,
-        new Fr(lowLeafPreimage!.getNextKey()),
-        lowLeafPreimage!.getNextIndex(),
+    const leafSlot = computePublicDataTreeLeafSlot(contractAddress, slot);
+
+    if (this.doMerkleOperations) {
+      // Get leaf if present, low leaf if absent
+      // If leaf is present, hint/trace it. Otherwise, hint/trace the low leaf.
+      const [leafIndex, leafPreimage, leafPath, _alreadyPresent] = await getLeafOrLowLeaf<PublicDataTreeLeafPreimage>(
+        MerkleTreeId.PUBLIC_DATA_TREE,
+        leafSlot.toBigInt(),
+        this.merkleTrees,
       );
```

Comment:
I'm basically saying "if leaf is present, trace that. Otherwise trace the low leaf." Circuit will have to recognize that the hinted `leafPreimage.slot` does not match the computed `leafSlot`, confirm that the provided leaf is a low-leaf that skips `leafSlot` and then prove the low-leaf's membership and return the value 0 to the user. Does this make sense? I think we were hoping to always just hint the low-leaf, but for the public data tree, the low-leaf doesn't contain the storage value.

Reply:
yeah this makes sense, it's a good saving for the public data writes since we then still only hint 1 sibling path for everything.